### ISSUE TITLE
Corsa-e: Charge control not working

### ISSUE THREAD

**Issue by sOckhamSter** [flobz/psa_car_controller#137]
I'm not sure if this is a bug, or whether it's just because the (Vauxhall) Corsa-e doesn't support these features. I'm unable to get any of the charge control commands to work. The API appears to send them fine, they just don't do anything. so for example, sending the command:
http://localhost:5000/charge_now/YOURVIN/0
... does not stop the charge. Equally, if I set a time for the charge to stop, or set the percentage threshold, the car ignores them and carries on charging regardless.

Do you know if this is just because the car doesn't support this, or could something be up? The MyVauxhall app does *not* allow me to perform such actions which is why I suspect that the commands are not supposed to do anything.

**Comment by thommierother** [comment]
I don't think its a problem of the car itself. The corsa has the same technical platform as the PSA models and even the MyPeugeot App has no option to stop charging from remote (I do this from the wallbox). As far as I understand it, the car receives mqtt messages from the API, e.g. for starting the preconditioning. But other options which are already visible in the API are simply not implemented in the official mobile apps yet.

**Comment by flobz** [comment]
Look at the log.
You can retry the command after 30 second, sometime it's needed to reconnect.

**Comment by sOckhamSter** [comment]
I've enabled debug on server.py to see what's happening. When I send /charge_now/vin/0 to stop the charge, this is what appears in the logs:

```
2021-05-24 10:56:49,325 :: DEBUG :: vin:VXKUHZKXZLxxxxxxx longitude:None latitude:None date:2021-05-24 10:32:40+00:00 mileage:2052.8 level:46.0 charge_date:2021-05-24 10:32:40+00:00 level_fuel:None moving:None
2021-05-24 10:56:49,328 :: DEBUG :: position already saved
2021-05-24 10:56:49,329 :: DEBUG :: No abrp token provided
2021-05-24 10:56:49,331 :: DEBUG :: level already stored
2021-05-24 10:56:49,332 :: DEBUG :: charging_status:InProgress 
2021-05-24 10:56:49,333 :: ERROR :: Can't get charge hour: PT30M
Traceback (most recent call last):
  File "/psa_car_controller/my_psacc.py", line 344, in __get_charge_hour
    hour_minute = re.findall(reg, hour_str)[0]
IndexError: list index out of range
```

**Comment by sOckhamSter** [comment]
I've just noticed that in 2.3.2 you updated the code to understand PT times, but I can't see that it understands "PT30M", which is currently my default scheduled start time of 00:30. Could it be because of that reason it's crashing for me do you think?

**Comment by flobz** [comment]
fixed with 0a8c1eab4292ed89da8e707088e86fbfac249635
You have an error:
172.17.0.1 - - [24/May/2021 11:40:52] "GET /charge_now/VXKUHZKXZLxxxxxx/0 HTTP/1.1" 200 -                          
2021-05-24 11:40:52,145 :: WARNING :: Disconnected with result code 1

**Comment by sOckhamSter** [comment]
Thanks @flobz any ideas if I can dig into any diagnostics anywhere to get more information on what that error is or what result code 1 means?

**Comment by flobz** [comment]
It happen when the token expire. What happens when you send the command twice?

**Comment by sOckhamSter** [comment]
The same thing. I've sent it multiple times with gaps of at least 30 seconds in between, trying different scenarios (i.e. stopping an existing charge, starting a new charge, starting while waiting on a deferred charge, setting a scheduled charge start time, etc). Very odd. Perhaps I need to play around more and collect data... it's just annoying when the family keeps using the car and I can't test all the time!

**Comment by flobz** [comment]
Does it works to stop the charge from android app ?

**Comment by sOckhamSter** [comment]
There is no such functionality provided by the MyVauxhall app which was part of my original question - I didn't know if functionality which is not provided by the app itself would work. Using the MyVauxhall app, all I can do is:
- Activate / Deactivate climate
- Schedule climate
- Schedule a deferred charge start time

**Comment by flobz** [comment]
when you stop the charge in psa_car_controller, it does a Schedule a deferred charge start time

**Comment by sOckhamSter** [comment]
Ok, so does that mean when you start the charge, it sets the deferred charge start time to 'now'?
Using the app itself works fine to set deferred charge times, and the car is happy to obey. It just doesn't seem to work in psa_car_controller for me :(

**Comment by flobz** [comment]
> it sets the deferred charge start time to 'now'?
No it just stop the charge or delay to next day.

Could you try to remove otp.bin ?

**Comment by sOckhamSter** [comment]
Tried that, ran through the SMS/PIN routine, and the same thing happens.

**Comment by sOckhamSter** [comment]
Right then, sorry it's taken a while but I've had to play around to test several scenarios. I think I sort of have it working now but it's very fiddly. Until I'd connected the charger to the car and put it into deferred charging mode using the app itself, the charge stop/start wouldn't work. It now works, most of the time.

But, in order to get that far I had to first confirm that setting a deferred charge time via the app worked. After looking at your code I noticed that the /charge_hour API command requires form submission which is difficult to do using a URL call. I tweaked my version as follows:

```
#@app.route('/charge_hour')
#def change_charge_hour():
#    return jsonify(myp.change_charge_hour(request.form['vin'], request.form['hour'], request.form['minute']))

@app.route('/charge_hour/<string:vin>/<int:hour>/<int:minute>')
def change_charge_hour(vin, hour, minute):
    return jsonify(myp.change_charge_hour(vin, hour, minute))
```

It might be worth amending views.py to allow a call to change_charge_hour using a single non-form standard URL call do you think? That would make it easy to call using a Home Assistant switch.

**Comment by flobz** [comment]
Look at the domoticz exemple, it's quite simple to use.

**Comment by sOckhamSter** [comment]
I think I’m missing something- the examples only show how to use charge_control (to set a stop time), not charge_hour to set a start time?

**Comment by flobz** [comment]
you just have to do a request like this:
http://127.0.0.1:5000/charge_hour?vin=xxx&hour=22&min=0
